File: 820-find-eventual-safe-states/find-eventual-safe-states.cpp
class Solution {
public:
    
    bool isCycleDFS(int u, vector<vector<int>>&adj, vector<int>&visited, vector<int>&inRecursion) {
        
        
        // if(visited[u] && inRecursion[u]) return false;
        
        visited[u] = true;
        
        inRecursion[u] = true;
        
        for(auto &v : adj[u]) {
            
           
                
            if (visited[v] == false && isCycleDFS(v, adj, visited, inRecursion)) return true;
            
            else if(inRecursion[v]  )return true;
            
        }
        
        inRecursion[u] = false;
        
        return false;
    }
    
    vector<int> eventualSafeNodes(vector<vector<int>>& adj) {
        
                
        vector<int> inRecursion(adj.size(), false);
        vector<int> visited(adj.size(), false);

        for(int i = 0; i < adj.size(); i++) {
            
            if(visited[i]) continue;
            
            isCycleDFS(i, adj, visited, inRecursion);
            
        }
        vector<int> safeNodes;


        for(int i = 0; i < adj.size(); i++ ) {

            if(inRecursion[i] == false) safeNodes.push_back(i);
        }

        return safeNodes;
    }
};

File: 820-find-eventual-safe-states/find-eventual-safe-states.js
/**
 * @param {number[][]} graph
 * @return {number[]}
 */
var eventualSafeNodes = function(graph) {
    
    const ans = new Array();

    const visited = new Array(graph.length).fill(false);

    const inR = new Array(graph.length).fill(false);

    const adj = Array.from({length: graph.length}, () => new Array())

    for(let i = 0; i < graph.length; i++) {

        for(let j = 0; j < graph[i].length; j++) {

            adj[graph[i][j]].push(i);
        }
    }

    const freq = new Array(graph.length).fill(0);


    for(const u of adj) {

        for(const  v of u) freq[v]++;
    }

    const queue = new Array();

    for(let i = 0; i < freq.length; i++) {

        if(freq[i] == 0) queue.push(i)
    }

    let i = 0;
    
    while(i < queue.length) {

        let u = queue[i++];
        ans.push(u);

        for(const v of adj[u]) {

            freq[v]--;

            if(freq[v] == 0) queue.push(v);
        }
    }

    return ans.sort((a, b) => a - b);

    



    console.log(freq)







    for(let i = 0 ; i < graph.length; i++) {

        if(visited[i]) continue;
        DFS(i);

    }

    for(let i = 0; i < inR.length; i++)  if(!inR[i])ans.push(i);
    

    function DFS(u) {

        visited[u] = true;
        inR[u] = true;

        for(const v of graph[u]) {

            if(visited[v] == false && DFS(v)) return true;

            else if(inR[v]) return true;
        }
        inR[u] = false;
        return false;
    }

    return ans;
    
};